### PR TITLE
feat(datasets): streaming read-back + HF Storage Bucket sync (physical-AI data loop)

### PR DESCRIPTION
**Pull request by cagataycali**
## Summary

Integrates the LeRobot **streaming data loop** and **Hugging Face Storage Buckets** into strands-robots dataset collection — the workflow from Steven Palma's *"The physical-AI data loop: LeRobot and Hugging Face Buckets"* (HF, Jun 2026).

Today strands-robots can **record** a `LeRobotDataset` (sim + hardware) and `push_to_hub()` to a versioned dataset repo, but it has **no streaming read-back** and **no bucket-write** path. This PR closes that loop natively:

- **Collect** → `sim.stop_recording(bucket=...)` dumps to a mutable, Xet-deduped HF Storage Bucket (Phase 1/2 — avoids git-LFS history bloat during collection).
- **Train/eval** → `sim.stream_dataset(...)` reads frames straight from the Hub with no full download (Phase 3). Camera frames decode on the fly from MP4 shards; state/action from parquet.
- **Zero-touch on macOS** → an import-time dyld shim makes torchcodec video decode work out of the box.

### Native API (no library internals leaking into user code)
- **`Robot().stream_dataset(repo_id, **kw)`** — in-process streaming read-back, the natural counterpart to `start_recording`/`stop_recording`. Lives on `RecordingMixin`; the same `Robot()` that records can stream.
- **`stop_recording(..., bucket=, run_id=)`** — one kwarg syncs the finished episode to `hf://buckets/...`. Finalizes first so `meta/` (stats/info) ships → downstream normalization survives.
- **`DatasetRecorder.sync_to_bucket(bucket, run_id=None)`** — the underlying mutable-bucket sync via the `hf` CLI.
- **`StreamingDatasetReader`** (`strands_robots.streaming_dataset`) — version-tolerant wrapper over lerobot's `StreamingLeRobotDataset` (lazy import, `**kwargs` introspection, delta-grid validation parity, `drop_videos` proprio-only path for torchcodec-less edge devices). Exported lazily from the package root.

### Dependencies (the `[lerobot]` extra is now self-sufficient for the data loop)
- `lerobot[feetech,dataset]` — pulls `datasets` + `pandas` + `pyarrow` + `av` + `torchcodec` so `StreamingLeRobotDataset` video decode works without users hand-installing decode deps.
- `torchcodec` pinned to track the resolved torch C++ ABI: `0.10` on darwin/arm64 (torch 2.10), `0.11` on linux/aarch64 (torch 2.11). Fixes the `dlopen "Symbol not found: c10::MessageLogger"` mismatch that 0.7/0.8 hit against torch 2.10.

### macOS zero-touch dyld shim (`strands_robots/_dyld.py`)
torchcodec links ffmpeg via `@rpath`, but Homebrew's ffmpeg (`/opt/homebrew/lib`) isn't on the default dyld search path → `import torchcodec` fails with `Library not loaded: @rpath/libavutil.59.dylib`. `os.environ` mid-process can't fix it (dyld snapshots `DYLD_*` at launch), and `ctypes` preload doesn't satisfy `@rpath` on macOS. So `ensure_ffmpeg_on_dyld_path()` runs eagerly at `import strands_robots`:
- No-op unless macOS + Homebrew ffmpeg present + torchcodec installed + var not already set.
- Exports `DYLD_FALLBACK_LIBRARY_PATH` so **child processes** (DataLoader workers, subprocess training) inherit it immediately.
- **Re-execs the interpreter once** (guarded against loops) for the current process — but **only when safe**: refuses inside Jupyter / IPython / pytest / REPL / `python -c`, where it warns with the one-line `export` instead of replacing the host process.
- Opt out with `STRANDS_ROBOTS_NO_DYLD_SHIM=1`.

### Example + docs
- `examples/06_agent_collect_and_stream.py` — a Strands `Agent` records a dataset from one NL prompt, then `sim.stream_dataset()` reads it back with full camera decode. ~50 lines, no torchcodec/av plumbing.
- README: new **"Recording & streaming datasets"** section + the macOS note + `STRANDS_ROBOTS_NO_DYLD_SHIM` env var.
- `docs/recording.md`: `sync_to_bucket` in the methods table + a **"Stream back (no full download)"** section.

## Verification (local, uv + py3.12, lerobot 0.5.2 from source, torch 2.10 + torchcodec 0.10, `MUJOCO_GL=cgl`)
- Agent-driven sim recording: 60 frames / 1 episode, 2 cameras, `libsvtav1`.
- `sim.stream_dataset()` yields `state(6,)` + `action(6,)` + **decoded** camera tensors `(3, 480, 640)` straight from MP4 shards — zero re-download.
- Zero-touch confirmed: `env -u DYLD_FALLBACK_LIBRARY_PATH python examples/06_*.py` streams video with no manual setup.
- **56 tests pass** (14 new streaming/shim/facade + 42 existing recorder). pytest correctly does **not** re-exec.

## Notes / scope
- For full **training**, the upstream trainer already uses the same engine via `python -m lerobot.scripts.train dataset.streaming=true` — no new code needed there; `StreamingDatasetReader`'s home is in-process eval/replay/inspection.
- True `hf://buckets/...` **streaming** (vs collection-time sync) needs a small upstream lerobot change to `LeRobotDatasetMetadata.url_root`; the recommended path today is collect→bucket + stream from a published dataset repo (near-instant server-side `hf buckets cp`). Tracked as a follow-up.
- Backward compatible: all additions; `stop_recording()`'s new args are keyword-only with prior defaults.

## Files
```
 strands_robots/streaming_dataset.py           | +199  (new)
 strands_robots/_dyld.py                        | +159  (new)
 strands_robots/dataset_recorder.py             |  +75  (sync_to_bucket)
 strands_robots/simulation/mujoco/recording.py  |  +81  (stream_dataset, bucket=)
 strands_robots/__init__.py                     |  +15  (lazy export + shim)
 pyproject.toml                                 |  +23  (lerobot[dataset] + torchcodec)
 examples/06_agent_collect_and_stream.py        |  +64  (new)
 tests/test_streaming_dataset.py                | +287  (new, 14 tests)
 README.md                                      |  +63
 docs/recording.md                              |  +54
```

---

## §13 Review Rounds

| Round | Concern | Fix commit | Pin test |
|-------|---------|------------|----------|
| R1 | LLM-input-safety: `sync_to_bucket` bucket/run_id unvalidated into subprocess | `058b11d` | `tests/test_streaming_dataset.py::test_sync_to_bucket_rejects_*` (12 parametrized cases) |
| R1 | CodeQL: `StreamingDatasetReader` exported but not defined | `058b11d` | (static analysis, no runtime test needed) |
| R1 | CI: feetech extra test assertion too strict for comma-separated extras | `ee7c093` + `0eb260f` (format) | `tests/test_customer_workflow_regressions.py::test_lerobot_extra_includes_feetech` |

---

🤖 Authored with assistance from DevDuck.

### CODE REVIEW

File: strands_robots/dataset_recorder.py [hunk 1]
```diff
@@ -631,6 +631,81 @@ def push_to_hub(
             logger.error("push_to_hub failed: %s", e)
             return {"status": "error", "message": str(e)}
 
+    def sync_to_bucket(
```
Comment (yinsong1986) [inline strands_robots/dataset_recorder.py:646]: [MUST FIX] `sync_to_bucket` interpolates `bucket` and `run_id` directly into `subprocess.run` argv (`["hf", "buckets", "create", bucket]`, `["hf", "sync", local_root, dest]`) and into the URI `dest = f"hf://buckets/{bucket}/{run_id}"` with no validation. These are agent-reachable: `stop_recording(bucket=, run_id=)` is dispatched through `Simulation._dispatch_action`, and `_validate_and_build_kwargs` (simulation.py:2439) only rejects unknown param names and checks vector lengths/required params — string params matching the method signature flow through unvalidated straight from the LLM-emitted action dict.

This is must-fix under AGENTS.md > `Review Learnings (PR #92 - CI Security Baseline)` > “LLM Input Safety”: *“every parameter on an agent-callable tool … that flows into `subprocess.run` … MUST be validated up front via regex allowlist … Argv-style subprocess does not exempt you - defense-in-depth”* and *“Reject … `..` path traversal components.”* `run_id` defaults to `self.dataset.repo_id.split("/")[-1]` and a value like `../other-bucket` or one containing shell metacharacters reaches the CLI/URI unchecked.

Resolution: validate `bucket` and `run_id` against an allowlist (e.g. `^[a-zA-Z0-9._-]+(/[a-zA-Z0-9._-]+)?$` for `bucket`, `^[a-zA-Z0-9._-]+$` for `run_id`) and reject `..` components before constructing `dest`/the argv, returning a `{"status": "error", ...}` dict on rejection — mirroring the centralized `validate_inputs(...)` pattern AGENTS.md cites from `gr00t_inference`.

**Review by yinsong1986 — COMMENTED** [review]
## Summary

This PR closes the LeRobot "physical-AI data loop" inside strands-robots: it adds streaming read-back (`StreamingDatasetReader` + `sim.stream_dataset(...)`), a mutable HF Storage Bucket sync (`DatasetRecorder.sync_to_bucket` + `stop_recording(bucket=...)`), and a macOS dyld shim so torchcodec finds Homebrew ffmpeg with zero user setup. Dependency bounds are tightened (`lerobot[feetech,dataset]`, platform-pinned torchcodec). The design discipline is good: lerobot/torchcodec are lazily imported, `StreamingDatasetReader.open` forwards kwargs via `inspect.signature` introspection for version tolerance, the recorder finalizes `meta/` before any bucket sync (so downstream normalization stats ship), `stop_recording`'s new args are keyword-only and backward compatible, and the `_dyld.py` re-exec is gated tightly (refuses inside pytest/Jupyter/REPL/`-c`, guarded against exec loops, opt-out env var). 14 new unit tests inject fakes so they run without lerobot or the `hf` CLI.

The one blocking concern is an LLM-input-safety gap on the new bucket-sync path, flagged inline.

## Verification suggestions

- Confirm the dispatch reachability: `Robot("so100")` -> `_dispatch_action("stop_recording", {"bucket": "x/..", "run_id": "../y"})` passes `_validate_and_build_kwargs` (string params are not allowlisted) and reaches `sync_to_bucket`, where `bucket`/`run_id` are interpolated into `hf://buckets/{bucket}/{run_id}` and into `subprocess.run` argv.
- `hatch run test tests/test_streaming_dataset.py` to confirm the 14 new tests pass without lerobot/`hf` installed.

**Comment by cagataycali** [comment]
Pushed 058b11d addressing both review threads.

**`sync_to_bucket` input validation (must-fix).** `bucket` and `run_id` are now allowlist-validated before any `subprocess.run` argv or `hf://buckets/...` URI interpolation, since `stop_recording(bucket=, run_id=)` is agent-reachable through `_dispatch_action` and string params matching the signature previously flowed through unchecked:

```python
_BUCKET_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*(/[A-Za-z0-9][A-Za-z0-9._-]*)?$")
_RUN_ID_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")
```

`bucket` is `name` or `org/name`; `run_id` is a single path segment. Both reject `..` traversal, shell metacharacters, embedded separators, and a leading dash, returning `{"status": "error", ...}` before `hf` runs - mirroring the centralized fail-closed pattern used by `gr00t_inference`. 12 parametrized regression cases assert the subprocess never runs for a rejected target; they fail on pre-fix code.

**CodeQL `StreamingDatasetReader` exported-but-not-defined.** Added the symbol to the `TYPE_CHECKING` import block alongside the other lazy exports, so static analysis sees it defined (runtime still resolves it via the PEP 562 `__getattr__` lazy map).

Also cleared the lint/type debt that failed `call-test-lint`: `Callable` now imports from `collections.abc` (UP035), the quoted self-type is unquoted (UP037), and `streaming_dataset.py` has full return/param annotations so mypy is clean on the module. Local gate green: `ruff check` / `ruff format --check` clean on `strands_robots tests tests_integ`, mypy clean on the changed files, and 69 recorder+streaming tests pass.

File: strands_robots/dataset_recorder.py (hunk 1 above)
Comment (cagataycali) [reply strands_robots/dataset_recorder.py:646]: Fixed in 058b11d. `bucket` and `run_id` are now validated against anchored allowlists before any subprocess or URI interpolation:

- `_BUCKET_RE = ^[A-Za-z0-9][A-Za-z0-9._-]*(/[A-Za-z0-9][A-Za-z0-9._-]*)?$` (`name` or `org/name`)
- `_RUN_ID_RE = ^[A-Za-z0-9][A-Za-z0-9._-]*$` (single path segment)

The leading-char anchor `[A-Za-z0-9]` rejects leading-dot traversal (`..`, `../escape`, `org/../escape`), and the charset excludes `/` (in run_id), spaces, and shell metacharacters. A rejected value returns `{"status": "error", ...}` before `dest` is built and before `hf` runs. `run_id`'s default (`repo_id.split("/")[-1]`) is validated on the same path.

Pinned regression tests in `tests/test_streaming_dataset.py`: `test_sync_to_bucket_rejects_unsafe_bucket` and `test_sync_to_bucket_rejects_unsafe_run_id` (parametrized over `../escape`, `org/../escape`, metacharacters, extra segments) assert subprocess is never invoked on rejection. 16 sync_to_bucket tests pass locally.

**Comment by cagataycali** [comment]
Superseded by #611 — a clean single-commit branch off the latest `main` (this PR accumulated merge commits from concurrent `main` updates). Same code, same 11 files; closing in favor of #611.
